### PR TITLE
Add caching to navigation and snippet controller

### PR DESCRIPTION
**Pull request by Prokyonn**
Add caching and tag invalidation support for `NavigationController` and `SnippetAreaController`

### CODE REVIEW

File: Controller/NavigationController.php [hunk 1]
```diff
@@ -28,6 +30,9 @@ class NavigationController
 {
     use RequestParametersTrait;
 
+    // 7 days
+    public const NAVIGATION_CACHE_LIFE_TIME = '604800';
```
Comment (alexander-schranz) [inline Controller/NavigationController.php]: Should be configureable.

File: Controller/NavigationController.php [hunk 2]
```diff
@@ -84,6 +98,16 @@ public function getAction(Request $request, string $context): Response
                 'Content-Type' => 'application/json',
             ]
         );
+        $response->setPublic();
+        $response->setMaxAge(0);
+        $response->setSharedMaxAge(0);
```
Comment (alexander-schranz) [inline Controller/NavigationController.php]: Should be the values from: https://github.com/sulu/sulu/blob/75578fc6074db7f31d63eb4e958b1e6aac676423/src/Sulu/Bundle/HttpCacheBundle/CacheLifetime/CacheLifetimeEnhancer.php#L77-L78

File: Controller/SnippetAreaController.php [hunk 3]
```diff
@@ -30,6 +32,9 @@ class SnippetAreaController
 {
     use RequestParametersTrait;
 
+    // 7 days
+    public const SNIPPET_AREA_CACHE_LIFE_TIME = '604800';
```
Comment (alexander-schranz) [inline Controller/SnippetAreaController.php]: Should be configureable.

File: Controller/SnippetAreaController.php [hunk 4]
```diff
@@ -113,5 +127,16 @@ public function getAction(Request $request, string $area): Response
                 'Content-Type' => 'application/json',
             ]
         );
+
+        $response->setPublic();
+        $response->setMaxAge(0);
+        $response->setSharedMaxAge(0);
```
Comment (alexander-schranz) [inline Controller/SnippetAreaController.php]: Should be the values from: https://github.com/sulu/sulu/blob/75578fc6074db7f31d63eb4e958b1e6aac676423/src/Sulu/Bundle/HttpCacheBundle/CacheLifetime/CacheLifetimeEnhancer.php#L77-L78

File: EventSubscriber/NavigationInvalidationSubscriber.php [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+use Sulu\Bundle\DocumentManagerBundle\Bridge\PropertyEncoder;
+use Sulu\Bundle\HttpCacheBundle\Cache\CacheManager;
+use Sulu\Component\DocumentManager\Event\PublishEvent;
+use Sulu\Component\DocumentManager\Event\RemoveEvent;
+use Sulu\Component\DocumentManager\Event\RemoveLocaleEvent;
+use Sulu\Component\DocumentManager\Event\UnpublishEvent;
+use Sulu\Component\DocumentManager\Events;
+use Symfony\Component\EventDispatcher\EventSubscriberInterface;
+
+class NavigationInvalidationSubscriber implements EventSubscriberInterface
```
Comment (alexander-schranz) [inline EventSubscriber/NavigationInvalidationSubscriber.php]:
```suggestion
/**
 * @final
 * @internal
 */
class NavigationInvalidationSubscriber implements EventSubscriberInterface
```

Should be kept internal and final.

**Review by alexander-schranz — CHANGES_REQUESTED** [review]
Missing a subscriber which invalidates snippet areas.

File: EventSubscriber/NavigationInvalidationSubscriber.php [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+        $this->liveSession = $liveSession;
+    }
+
+    public static function getSubscribedEvents()
+    {
+        return [
+            Events::PUBLISH => ['invalidateNavigationContextBeforePublishing', 8192],
+            Events::UNPUBLISH => ['invalidateNavigationContextBeforeUnpublishing', 8192],
+            Events::REMOVE => ['invalidateNavigationContextBeforeRemoving', 8192],
+            Events::REMOVE_LOCALE => ['invalidateNavigationContextBeforeRemovingLocale', 8192],
```
Comment (alexander-schranz) [inline EventSubscriber/NavigationInvalidationSubscriber.php]: This listeners should just collect the data.

And a `Flush` method with allow priority should then call the CacheManager.

Do not forget to implement `kernel.reset` event to empty the array and empty the array.

**Comment by Prokyonn** [comment]
> Missing a subscriber which invalidates snippet areas.

@alexander-schranz do we really need one?

At the moment it is working, as soon as the snippet is modified, the tag is invalidated through the default [InvalidationSubscriber](https://github.com/sulu/sulu/blob/2.5/src/Sulu/Bundle/HttpCacheBundle/EventSubscriber/InvalidationSubscriber.php#L126) from the HttpCacheBundle

**Comment by alexander-schranz** [comment]
@Prokyonn but if a default snippet is changed to another default snippet the cache need to be invalided also. Else you will get a long time the old default snippet.

File: EventSubscriber/NavigationInvalidationSubscriber.php [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+use Sulu\Component\DocumentManager\Event\UnpublishEvent;
+use Sulu\Component\DocumentManager\Events;
+use Symfony\Component\EventDispatcher\EventSubscriberInterface;
+
+/**
+ * @final
+ *
+ * @internal
+ */
+class NavigationInvalidationSubscriber implements EventSubscriberInterface
```
Comment (alexander-schranz) [inline EventSubscriber/NavigationInvalidationSubscriber.php]: would add the `ResetInterface` here.

**Review by alexander-schranz — APPROVED** [review]

**Pull request merged**